File: jung-api/src/main/java/edu/uci/ics/jung/graph/util/ParallelEdgeIndexFunction.java
/*
 * Created on Sep 24, 2005
 *
 * Copyright (c) 2005, The JUNG Authors
 *
 * All rights reserved.
 *
 * This software is open-source under the BSD license; see either
 * "license.txt" or
 * https://github.com/jrtom/jung/blob/master/LICENSE for a description.
 */
package edu.uci.ics.jung.graph.util;

import static com.google.common.base.Preconditions.checkNotNull;

import com.google.common.graph.EndpointPair;
import com.google.common.graph.Network;
import java.util.HashMap;
import java.util.Map;

/**
 * A class which creates and maintains indices for parallel edges. Parallel edges are defined here
 * to be the collection of edges that are returned by <code>graph.edgesConnecting(v, w)</code> for
 * some <code>v</code> and <code>w</code>.
 *
 * <p>At this time, users are responsible for resetting the indices (by calling <code>reset()</code>
 * ) if changes to the graph make it appropriate.
 *
 * @author Joshua O'Madadhain
 * @author Tom Nelson
 */
public class ParallelEdgeIndexFunction<N, E> implements EdgeIndexFunction<E> {
  protected Map<E, Integer> edgeIndex = new HashMap<E, Integer>();
  protected Network<N, E> graph;

  /**
   * @param graph the graph for which this index function is defined
   */
  public ParallelEdgeIndexFunction(Network<N, E> graph) {
    this.graph = checkNotNull(graph, "graph");
  }

  public int getIndex(E edge) {
    checkNotNull(edge, "edge");
    Integer index = edgeIndex.get(edge);
    if (index == null) {
      EndpointPair<N> endpoints = graph.incidentNodes(edge);
      N u = endpoints.nodeU();
      N v = endpoints.nodeV();
      int count = 0;
      for (E connectingEdge : graph.edgesConnecting(u, v)) {
        edgeIndex.put(connectingEdge, count++);
      }
      return edgeIndex.get(edge);
    }
    return index;
  }

  public void reset(E edge) {
    edgeIndex.remove(checkNotNull(edge, "edge"));
  }

  public void reset() {
    edgeIndex.clear();
  }
}

File: jung-api/src/main/kotlin/edu/uci/ics/jung/graph/util/ParallelEdgeIndexFunction.kt
/*
 * Created on Sep 24, 2005
 *
 * Copyright (c) 2005, The JUNG Authors
 *
 * All rights reserved.
 *
 * This software is open-source under the BSD license; see either
 * "license.txt" or
 * https://github.com/jrtom/jung/blob/master/LICENSE for a description.
 */
package edu.uci.ics.jung.graph.util

import com.google.common.base.Preconditions.checkNotNull
import com.google.common.graph.Network

/**
 * A class which creates and maintains indices for parallel edges. Parallel edges are defined here
 * to be the collection of edges that are returned by `graph.edgesConnecting(v, w)` for
 * some `v` and `w`.
 *
 * At this time, users are responsible for resetting the indices (by calling `reset()`)
 * if changes to the graph make it appropriate.
 *
 * @author Joshua O'Madadhain
 * @author Tom Nelson
 */
class ParallelEdgeIndexFunction<N : Any, E : Any>(
  private val graph: Network<N, E>
) : EdgeIndexFunction<E> {

  private val edgeIndex: MutableMap<E, Int> = HashMap()

  init {
    checkNotNull(graph, "graph")
  }

  /**
   * @param graph the graph for which this index function is defined
   */

  override fun getIndex(edge: E): Int {
    checkNotNull(edge, "edge")
    val index = edgeIndex[edge]
    if (index == null) {
      val endpoints = graph.incidentNodes(edge)
      val u = endpoints.nodeU()
      val v = endpoints.nodeV()
      var count = 0
      for (connectingEdge in graph.edgesConnecting(u, v)) {
        edgeIndex[connectingEdge] = count++
      }
      return edgeIndex[edge]!!
    }
    return index
  }

  override fun reset(edge: E) {
    edgeIndex.remove(checkNotNull(edge, "edge"))
  }

  override fun reset() {
    edgeIndex.clear()
  }
}
